File: leetcode/2105/Solution.py
from typing import List

class Solution:
    def minimumRefill(self, plants: List[int], capacityA: int, capacityB: int) -> int:
        n = len(plants)
        numOfRefill = 0
        l = 0
        r = n - 1
        curA = capacityA
        curB = capacityB
        while l < r:
            if curA < plants[l]:
                curA = capacityA
                numOfRefill += 1
            curA -= plants[l]
            if curB < plants[r]:
                curB = capacityB
                numOfRefill += 1
            curB -= plants[r]
            l += 1
            r -= 1
        if l == r and max(curA, curB) < plants[l]:
            numOfRefill += 1
        return numOfRefill

if __name__ == '__main__':
    sol = Solution()
    plants = [2,2,3,3]
    capacityA = 3
    capacityB = 4
    print("test:", sol.minimumRefill(plants, capacityA, capacityB))

File: leetcode/2105/Solution.java
class Solution {
    public int minimumRefill(int[] plants, int capacityA, int capacityB) {
        int n = plants.length;
        int numOfRefill = 0;
        int l = 0;
        int r = n - 1;
        int curA = capacityA;
        int curB = capacityB;
        while (l < r) {
            if (curA < plants[l]) {
                curA = capacityA;
                numOfRefill++;
            }
            curA -= plants[l];
            if (curB < plants[r]) {
                curB = capacityB;
                numOfRefill++;
            }
            curB -= plants[r];
            l++;
            r--;
        }
        if (l == r) {
            // they reach then same plant
            if (Math.max(curA, curB) < plants[l]) {
                // need refill one more time
                numOfRefill++;
            }
        }
        return numOfRefill;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] plants = {2,2,3,3};
        int capacityA = 3;
        int capacityB = 4;
        System.out.println("test: " + sol.minimumRefill(plants, capacityA, capacityB));
    }
}